extensions/amp-next-page/0.2/service.js

```diff
@@ -0,0 +1,476 @@
+    if (this.injectedPositionObserver_) {
+      return this.injectedPositionObserver_;
+    }
+
+    installPositionObserverServiceForDoc(this.ampdoc_);
+    return Services.positionObserverForDoc(this.ampdoc_.getHeadNode());
+  }
+
+  /**
+   * @param {!Page} page
+   */
+  setLastFetchedPage(page) {
+    this.lastFetchedPage_ = page;
+    this.printPagesForTesting();
+  }
+
+  /**
+   *
+   * @param {!Page} page
+   * @param {!Document} doc
+   * @return {?../../../src/runtime.ShadowDoc}
+   */
+  appendAndObservePage(page, doc) {
+    // If the user already scrolled to the bottom, prevent rendering
+    if (this.getViewportsAway_() >= 1) {
+      // TODO(wassgha): Append a "load next article" button
+      return null;
+    }
+
+    this.element_.insertBefore(this.separator_.cloneNode(true), this.moreBox_);
+
+    const header = this.win_.document.createElement('div');
+    const shadowRoot = this.win_.document.createElement('div');
+    const footer = this.win_.document.createElement('div');
+
+    this.element_.insertBefore(header, this.moreBox_);
+    this.element_.insertBefore(shadowRoot, this.moreBox_);
+    this.element_.insertBefore(footer, this.moreBox_);
+
+    // TODO(wassgha): Unobserve
+    this.getPositionObserver_().observe(
+      header,
+      PositionObserverFidelity.LOW,
+      page.boundPositionChanged.bind(page, PageBound.HEADER)
+    );
+    this.getPositionObserver_().observe(
+      footer,
+      PositionObserverFidelity.LOW,
+      page.boundPositionChanged.bind(page, PageBound.FOOTER)
+    );
+
+    // Handles extension deny-lists and sticky items
+    sanitizeDoc(doc);
+
+    // Try inserting the shadow document
+    try {
+      const amp = this.win_.AMP.attachShadowDoc(shadowRoot, doc, '', {
+        visibilityState: VisibilityState.PRERENDER,
+      });
+
+      const ampdoc = devAssert(amp.ampdoc);
+      installStylesForDoc(ampdoc, CSS, null, false, TAG);
+
+      const body = ampdoc.getBody();
+      body.classList.add('i-amphtml-next-page-document');
+
+      return amp;
+    } catch (e) {
+      dev().error(TAG, 'failed to attach shadow document for page', e);
+      return null;
+    }
+  }
+
+  /**
+   * @return {number} viewports left to reach the end of the document
+   * @private
+   */
+  getViewportsAway_() {
+    return Math.round(
+      (this.viewport_.getScrollHeight() -
+        this.viewport_.getScrollTop() -
+        this.viewport_.getHeight()) /
+        this.viewport_.getHeight()
+    );
+  }
+
+  /**
+   * @private
+   */
+  updateScrollDirection_() {
+    const scrollTop = this.viewport_.getScrollTop();
+    this.scrollDirection_ =
+      scrollTop > this.lastScrollTop_ ? Direction.DOWN : Direction.UP;
+    this.lastScrollTop_ = scrollTop;
+  }
+
+  /**
+   * @param {Page} desiredPage
+   * @return {number} The index of the page.
+   */
+  getPageIndex_(desiredPage) {
+    const pages = dev().assertArray(this.pages_);
+    return findIndex(pages, page => page === desiredPage);
```

Comment:
Since we are using a `===` predicate, this is equivalent to `pages.indexOf(desiredPage)`, so I think we should use that instead, since all browsers support it.